src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureHandler.kt

```diff
@@ -16,15 +16,26 @@ import com.intellij.refactoring.RefactoringBundle
 import com.intellij.refactoring.changeSignature.ChangeSignatureHandler
 import com.intellij.refactoring.util.CommonRefactoringUtil
 import org.rust.RsBundle
-import org.rust.lang.core.psi.RsFunction
+import org.rust.lang.core.psi.*
 import org.rust.lang.core.psi.ext.*
 import org.rust.openapiext.editor
 import org.rust.openapiext.elementUnderCaretInEditor
 
 class RsChangeSignatureHandler : ChangeSignatureHandler {
     override fun getTargetNotFoundMessage(): String = "The caret should be positioned at a function or method"
 
-    override fun findTargetMember(element: PsiElement): RsFunction? = element.ancestorOrSelf()
+    override fun findTargetMember(element: PsiElement): RsFunction? {
+        val function = element.ancestorOrSelf<RsFunction>(stopAt = RsBlock::class.java)
+        if (function is RsFunction) return function
+
+        return element.ancestors.mapNotNull { el ->
+            when (el) {
+                is RsCallExpr -> (el.expr as? RsPathExpr)?.path?.reference?.resolve() as? RsFunction
+                is RsMethodCall -> el.reference.resolve() as? RsFunction
+                else -> null
+            }
+        }.firstOrNull()
```

Comment:
This code works as expected for valid code when the plugin is able to provide proper name resolution. But if the plugin failed to resolve function/method call, we will try to find parent call that's quite unexpected.

For example,
```rust
fn main() {
    bar(foo/*caret*/(1));
}

fn bar(x: i32) {}
```

I expect that IDE shows me some warning (because `foo` doesn't exist and cannot be refactored) but currently we start to refactor `bar`

Reply:
I suppose we can fix this issue and generalize current implementation in the following way:
```kotlin
override fun findTargetMember(element: PsiElement): RsFunction? {
    for (el in element.ancestors) {
        return when (el) {
            is RsFunction -> el
            is RsCallExpr -> (el.expr as? RsPathExpr)?.path?.reference?.resolve() as? RsFunction
            is RsMethodCall -> el.reference.resolve() as? RsFunction
            is RsBlock -> return null
            else -> continue
        }
    }
    return null
}
```

Reply:
Good point. I changed the implementation and added a corresponding test.